### PR TITLE
Bump workflow (py-morphology-repair-workflow 2.0.3)

### PR DESCRIPTION
**Pull request by GianlucaFicarelli**
Replace https://github.com/BlueBrain/spack/pull/1058

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-morph-tool/package.py
Comment (asanin-epfl) [inline var/spack/repos/builtin/packages/py-morph-tool/package.py]: You mentioned that would be good to constrain it by upper bound, no? `py-neurom@1.8.0:1.999`?

File: var/spack/repos/builtin/packages/py-morph-tool/package.py
Comment (asanin-epfl) [inline var/spack/repos/builtin/packages/py-morph-tool/package.py:31]: Might the upper constraint bite us as dask switches to calendar versioning since 2.30?

File: var/spack/repos/builtin/packages/py-morph-tool/package.py
Comment (GianlucaFicarelli) [reply var/spack/repos/builtin/packages/py-morph-tool/package.py]: Right, but the CI is still failing and I didn't want to add more reasons to fail, so I'll need to add the constraints later.
However if I'm not wrong `py-morphology-repair-workflow 2.0.3` requires `neurom 2`, so all the other packages required by `py-morphology-repair-workflow` should be compatible with `neurom 2`.
If this is not the case, we should fix them before merging this PR.

File: var/spack/repos/builtin/packages/py-morph-tool/package.py
Comment (GianlucaFicarelli) [reply var/spack/repos/builtin/packages/py-morph-tool/package.py:31]: Yes, I think it's a good (safer) idea. We should update also `setup.py` in the package repository, though (at the moment it's `dask[bag]>=2.19.0` in the current master at https://github.com/BlueBrain/morph-tool/blob/1dc791e34dc49e3afaf3b689b277b4f6ea774e44/setup.py#L26).

File: var/spack/repos/builtin/packages/py-morph-tool/package.py
Comment (asanin-epfl) [reply var/spack/repos/builtin/packages/py-morph-tool/package.py]: ok

File: var/spack/repos/builtin/packages/py-morph-tool/package.py
Comment (asanin-epfl) [reply var/spack/repos/builtin/packages/py-morph-tool/package.py:31]: ok, anyway it would be better if PR is approved by someone from the spack team. All good from my side.

**Comment by mgeplf** [comment]
LGTM; looks like another mpi variant issue blocking this?

**Comment by GianlucaFicarelli** [comment]
> LGTM; looks like another mpi variant issue blocking this?

I'm holding this PR waiting for https://github.com/BlueBrain/spack/pull/1100, that should fix this concretization error.
As mentioned in the previous comments I was thinking to pin `neurom<2` in the spack packages requiring `neurom` but not required by `py-morphology-repair-workflow` (i.e. excluding `py-morph-tool` that needs to support also `neurom>=2`).

**Comment by GianlucaFicarelli** [comment]
The CI is passing now, but I moved the PR to draft because the following packages are required by `py-morphology-repair-workflow` but they probably need to be updated to be compatible with `neurom >= 2` (i.e. the `neurom` version cannot be pinned to `< 2`):
```
py-bluepy
py-bluepysnap
py-morph-tool
py-morph-validator
```
These packages instead are not required and they could be pinned:
```
py-diameter-synthesis
py-tns
```
Alternatively, I could check if it's possible to make a transitory release in spack (not `2.x`, but something like `wf.0.1`, `wf.0.2` or any better non-numeric naming) to be used only with `py-morphology-repair-workflow` for the following packages (however, it's possible that the packages will not work correctly if they use any of the features changed or removed in `neruom`):
```
py-neurom
py-bluepy
py-bluepysnap
py-morph-tool
py-morph-validator
```

File: var/spack/repos/builtin/packages/py-morphology-repair-workflow/package.py
Comment (matz-e) [inline var/spack/repos/builtin/packages/py-morphology-repair-workflow/package.py:23]: Is this actually distributed or could this be done via multiprocessing?

**Review by matz-e — DISMISSED** [review]

File: var/spack/repos/builtin/packages/py-morphology-repair-workflow/package.py
Comment (GianlucaFicarelli) [reply var/spack/repos/builtin/packages/py-morphology-repair-workflow/package.py:23]: From what I can see `dask` in `morphology-repair-workflow` is used only in 
```
ok, failures = zip(*dask.bag.from_sequence(iterable).map(wrapped).compute())
```
There is no multi-node setup, but It should wrap multiprocessing.
Do you see any reason to avoid dask in this case? (In some other cases we used `joblib`)

**Comment by GianlucaFicarelli** [comment]
Needs https://bbpteam.epfl.ch/project/issues/browse/NSETM-1442 before merging.

File: var/spack/repos/builtin/packages/py-morphology-repair-workflow/package.py
Comment (matz-e) [reply var/spack/repos/builtin/packages/py-morphology-repair-workflow/package.py:23]: IMO, the question should more be why Dask and not just pure `multiprocessing`? Why add a complex dependency when `multiprocessing` could do it?

Same for `joblib`, if it's just to run stuff on a single node.

Comment (arnaudon) [reply var/spack/repos/builtin/packages/py-morphology-repair-workflow/package.py:23]: +1 for multiprocessing if it is not distributed. I have a PR open on that code to fix dependencies with latest versions of other codes, I can remove dask dependency in the same time.

**Comment by GianlucaFicarelli** [comment]
Retest this please

**Comment by GianlucaFicarelli** [comment]
There are errors in Software Tests / Neurodamus manual building at  https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/hpc.module-testing/detail/hpc.module-testing/3480/pipeline

@matz-e do you know if it can be something not related to this PR?

**Comment by GianlucaFicarelli** [comment]
> There are errors in Software Tests / Neurodamus manual building at https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/hpc.module-testing/detail/hpc.module-testing/3480/pipeline
> 
> @matz-e do you know if it can be something not related to this PR?

@matz-e or @pramodk , after comparing the log with another successful execution, I see that jenkins for this PR is using an old version of neurodamus-neocortex:
/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/modules/applications/2021-03/tcl/linux-rhel7-x86_64/neurodamus-neocortex/1.3-3.2.0 
instead of
/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/modules/applications/2021-04/tcl/linux-rhel7-x86_64/neurodamus-neocortex/1.4-3.2.2

I think it could be fixed just cleaning the workspace /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/pulls/1099
If this is the case, what is the right way to delete it?

**Comment by matz-e** [comment]
> I think it could be fixed just cleaning the workspace /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/pulls/1099
> If this is the case, what is the right way to delete it?

Secret stuff by us :)

**Comment by matz-e** [comment]
How about a rebase to retest this? Working directory should be clean now.

**Comment by GianlucaFicarelli** [comment]
I rebased and pushed, but there is a new error at https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/hpc.module-testing/detail/hpc.module-testing/3485/pipeline:
```
/jenkins/10/workspace/hpc.module-testing@tmp/durable-a6834299/script.sh:
line 5: /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/pulls/1099/config/modules.sh: No such file or directory
```
Effectively this file is missing, how it should be generated?

**Comment by pramodk** [comment]
> Effectively this file is missing, how it should be generated?

thats typically misleading error. If you check & click on the `Test` tab then you will see that there are build errors (and hence module.sh is not generated):

image

**Comment by matz-e** [comment]
> thats typically misleading error

We could potentially handle this better, but every time I look into issues like this, I'm limited by our outdated Jenkins.

File: var/spack/repos/builtin/packages/brainbuilder/package.py
Comment (matz-e) [inline var/spack/repos/builtin/packages/brainbuilder/package.py:23]: Wow, you just keep bumping the dependencies ;)

File: var/spack/repos/builtin/packages/py-bluepy/package.py
Comment (matz-e) [inline var/spack/repos/builtin/packages/py-bluepy/package.py:45]: Could we in the future leave the obvious comments out?

File: var/spack/repos/builtin/packages/py-minis-validation/package.py
Comment (matz-e) [inline var/spack/repos/builtin/packages/py-minis-validation/package.py:21]: Is Pandas 2 going to break things?

**Review by matz-e — DISMISSED** [review]
Fine with me.

I remain concerned about all these generic upper limits to enforce the idea of semantic versioning. To me, this seems to assume that something is guaranteed to break, which may not necessarily be the case, but may very well cause us some concretization issues down the road.

**Comment by GianlucaFicarelli** [comment]
Sorry for the new changes, but after re-checking the packages I found that some packages were updated since the previous version. This should be the final update if the build is still successful.

Some notes about morphio:
I removed the old versions of morphio defined in spack because they were built from git, and the repo that was used for the submodule lexertl is no longer accessible, so the build was failing.

This is also causing the main spack build to fail on the default branch, and the latest versions of py-morphio/py-bluepy/py-bluepysnap today are not available in unstable.
A new repo for the submodule has been already set in the master branch of morphio, but a new version has not been released yet.

Changes related to the morphio versions:
- keeping the latest `3.1.1` (because the tar.gz including the submodules is available on pypi, and now it's used during the build)
- removing `3.0.2 2.7.1` (even though they could be kept because the tar.gz is available on pypi. Is there any reason to keep them?
There are several packages depending on morphio, but none of them is pinning morphio < 3 and >= 2.7.1 so they would not be used as dependencies, and py-morphio/2.7.1 can still be loaded from archive/2021-04).
- removing `2.3.10 2.3.4 2.2.1 2.1.2 2.0.8` (because the tar.gz is not available on pypi)

Moreover, I kept `morpho-kit 0.2.0` but I removed `morpho-kit 0.1.0` because it was depending on `morphio<=2.2.1`.

File: var/spack/repos/builtin/packages/py-bluepy/package.py
Comment (GianlucaFicarelli) [reply var/spack/repos/builtin/packages/py-bluepy/package.py:45]: Sure, I think that the bluepy spec will be cleaned when the old versions will be removed.

File: var/spack/repos/builtin/packages/py-minis-validation/package.py
Comment (GianlucaFicarelli) [reply var/spack/repos/builtin/packages/py-minis-validation/package.py:21]: I'm not sure, because I didn't work on the package, but I enforced the same constraints indicated in `setup.py` for consistency. Maybe it can be discussed outside of this PR and @asanin-epfl can add more comments about that.

File: var/spack/repos/builtin/packages/py-setuptools-scm/package.py
Comment (GianlucaFicarelli) [inline var/spack/repos/builtin/packages/py-setuptools-scm/package.py:22]: I think it's safe to use the `py-setuptools-scm+toml` by default because it's the preferred way to configure it.
I needed to change it to resolve a concretization issue.

File: var/spack/repos/builtin/packages/py-tqdm/package.py
Comment (GianlucaFicarelli) [inline var/spack/repos/builtin/packages/py-tqdm/package.py:27]: This is mostly from upstream.

File: var/spack/repos/builtin/packages/py-neurom/package.py
Comment (asanin-epfl) [inline var/spack/repos/builtin/packages/py-neurom/package.py]: This line almost repeats the line 42, can it be merged with it?

File: var/spack/repos/builtin/packages/py-minis-validation/package.py
Comment (asanin-epfl) [reply var/spack/repos/builtin/packages/py-minis-validation/package.py:21]: I've added upper constraints for additional safety. The reason for that was the Dask change to calendar versioning.

File: var/spack/repos/builtin/packages/py-neurom/package.py
Comment (GianlucaFicarelli) [reply var/spack/repos/builtin/packages/py-neurom/package.py]: I did in this way because it seemed clearer to me to split the dependencies based on the `py-neurom` version.
However, I'll try to merge `py-pandas` so we'll have one line less.

File: var/spack/repos/builtin/packages/py-minis-validation/package.py
Comment (GianlucaFicarelli) [reply var/spack/repos/builtin/packages/py-minis-validation/package.py:21]: In this case the constraint is for pandas and numpy, not dask (that's the one usually discussed 😄 ).

Comment (asanin-epfl) [reply var/spack/repos/builtin/packages/py-minis-validation/package.py:21]: yes, in order to be consistent among version constraints, I've applied this rule to all dependencies. I'm trying to follow this rule now everywhere as I find it useful to prevent unexpected surprises.

**Review by asanin-epfl — APPROVED** [review]
Looks good to me. As Matthias already approved it before, I'm approving.

File: var/spack/repos/builtin/packages/py-bluepy/package.py
Comment (tomdele) [reply var/spack/repos/builtin/packages/py-bluepy/package.py:45]: indeed. Not sure if a package still use version of bluepy<2.0.0.

**Review by tomdele — APPROVED** [review]
If CI is happy I am good

**Comment by GianlucaFicarelli** [comment]
Thanks all, I'm going to merge! If anyone is still reviewing, please comment and I'll open a separate PR.

**Comment by asanin-epfl** [comment]
Thank you for finishing this.

**Pull request merged**
